ui/src/actions/login.js

```diff
@@ -127,63 +90,56 @@ export function signIn(proposal, password, navigate) {
       serverIO.connect();
     }
 
-    fetch('mxcube/api/v0.1/login/', {
-      method: 'POST',
-      headers: {
-        Accept: 'application/json',
-        'Content-type': 'application/json',
-      },
-      credentials: 'include',
-      body: JSON.stringify({ proposal, password, previousUser }),
-    })
-      .then((response) => response.json())
-      .then(
-        (res) => {
-          if (res.code === 'ok') {
-            dispatch(showErrorPanel(false));
-            dispatch(getLoginInfo())
-              .then((response) => response)
-              .then((resp) => {
-                if (resp.loginType === 'User') {
-                  if (resp.user.inControl) {
-                    dispatch(showProposalsForm());
-                  } else {
-                    navigate('/');
-                  }
+    logIn(proposal, password, previousUser).then(
+      (res) => {
+        if (res.code === 'ok') {
+          dispatch(showErrorPanel(false));
+          dispatch(getLoginInfo())
+            .then((response) => response)
+            .then((resp) => {
+              if (resp.loginType === 'User') {
+                if (resp.user.inControl) {
+                  dispatch(showProposalsForm());
```

Comment:
Question:
How come we are so deeply nested?
Wouldn't it be better with async/await? Is it something that has been considered already?

Reply:
Yes, definitely on the radar but I'd like to first go over the login logic and move it out of Redux if possible.

Reply:
I think its a good idea as @axelboc said to do this in several steps to try to separate the fetching from the redux bit as much as we can and introducing an abstraction over `fetch` . The idea is then indeed, as we also said during the code camp, to start using `async` and `await`. I think the main reason that we started with the `then` for handling promises was that `async` and `await` was not widely adopted when the project was started. Then we just continued on that path.

Reply:
OK, of course, makes sense :)